### PR TITLE
enable ring0/pagetables to support arm64

### PR DESCRIPTION
**Pull request by lubinszARM**
Signed-off-by: Bin Lu <bin.lu@arm.com>

### CODE REVIEW

**Comment by lubinszARM** [comment]
Hi @amscanne 
The following test cases has been passed on Arm64 platform:
TestKernelSyscall
TestKernelFault
TestApplicationSyscall
TestApplicationFault
TestRegistersSyscall
TestInvalidate
TestEmptyAddressSpace
TestWrongVCPU
I will deliver the patches about kvm platform for Arm64 one by one.
And this is a basic patch for Arm64.

**Comment by lubinszARM** [comment]
image

For kvm platform on Arm64, there are 3 key scenarios in KVM, and 2 scenarios in host.
And there are 3 stacks in kvm, and 2 stacks in host.
Please see my design picture as reference.

File: pkg/sentry/platform/ring0/pagetables/pagetables_aarch64.go [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+	cont          = 0x1 << 52
+	pxn           = 0x1 << 53
+	xn            = 0x1 << 54
+	dirty         = 0x1 << 55
+	nG            = 0x1 << 11
+	shared        = 0x3 << 8
+)
+
+const (
+	mt_device_nGnRE = 0x1 << 2
```
Comment (avagin) [inline pkg/sentry/platform/ring0/pagetables/pagetables_aarch64.go]: This one is unused.

File: pkg/sentry/platform/ring0/pagetables/pagetables_aarch64.go [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+		// This is not allowed.
+		panic("SetSect called on valid page!")
+	}
+	atomic.StoreUintptr((*uintptr)(p), type_sect)
+}
+
+// IsSect returns true iff this page is a sect page.
+//
+//go:nosplit
+
```
Comment (avagin) [inline pkg/sentry/platform/ring0/pagetables/pagetables_aarch64.go]: unnecessary blank line?

File: pkg/sentry/platform/ring0/pagetables/pagetables_aarch64.go [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+		User:   v&user != 0,
+	}
+}
+
+// SetSect sets this page as a sect page.
+//
+// The page must not be valid or a panic will result.
+//
+//go:nosplit
+
```
Comment (avagin) [inline pkg/sentry/platform/ring0/pagetables/pagetables_aarch64.go]: unnecessary blank line?

File: pkg/sentry/platform/ring0/pagetables/pagetables_aarch64.go (hunk 1 above)
Comment (lubinszARM) [reply pkg/sentry/platform/ring0/pagetables/pagetables_aarch64.go]: Done.

File: pkg/sentry/platform/ring0/pagetables/pagetables_aarch64.go (hunk 2 above)
Comment (lubinszARM) [reply pkg/sentry/platform/ring0/pagetables/pagetables_aarch64.go]: Done.

File: pkg/sentry/platform/ring0/pagetables/pagetables_aarch64.go (hunk 3 above)
Comment (lubinszARM) [reply pkg/sentry/platform/ring0/pagetables/pagetables_aarch64.go]: Done.

**Comment by avagin** [comment]
Merged: https://github.com/google/gvisor/commit/07635d20d40e1a279c4b063abaaad51048400ed7

**Pull request closed without merging**